### PR TITLE
LNT ECMP automation test case

### PR DESCRIPTION
**Pull request by sjwahub**
Signed-off-by: Hiren Thakkar <hiren.thakkar@intel.com>

This PR  is to cover test cases automation for Linux networking ECMP on DPDK target. 
In addition, besides adding new entities in Makefile.am, reorg few existing entities. 

Test logs:
1. Test suite log run with ECMP test only  [ptf_ecmp_tests_run.log](https://github.com/ipdk-io/ovs/files/9439165/ptf_ecmp_tests_run.log)

2. Test suite log run with mixed all test cases including ECMP  [ptf_tests_runAll1.log](https://github.com/ipdk-io/ovs/files/9439167/ptf_tests_runAll1.log)

3. rebuilding OVS log, named "ovsbuild.log"  [ovsbuil.log](https://github.com/ipdk-io/ovs/files/9439169/ovsbuil.log)

### CODE REVIEW

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-1/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/3dddab33bdb5a12fe764aa4bc8e62e5fbd90b9a0

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-2/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/8780f93f92c9562be253101caba694247ae00f23

File: ptf_tests/common/lib/port_config.py [hunk 1]
```diff
@@ -273,11 +273,81 @@ def ipnetns_delete_namespace(self, namespace_name):
...
+            """
+            cmd = self.form_cmd(f" route del {dst}")
+            output, return_code, err = self.connection.execute_command(cmd)
+            if return_code:
+                raise ExecuteCMDException(f'FAIL:command "{cmd}" failed with '
+                                          f'an error {err}')
+            print(f"PASS: {cmd}")
+            return True
+        
+        def exec_general_cmd(self, cmd):
```
Comment (dkaranjai) [inline ptf_tests/common/lib/port_config.py]: No need for this in port_config.py as its a generic function. You can have this one in util itself, no need for a lib function for this

**Review by dkaranjai — CHANGES_REQUESTED** [review]

File: ptf_tests/common/lib/port_config.py (hunk 1 above)
Comment (sjwahub) [reply ptf_tests/common/lib/port_config.py]: Fixed it . Unit test log attached [Unittest.txt](https://github.com/ipdk-io/ovs/files/9443207/Unittest.txt)

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-3/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/43581ad28022b4c30e56691202e7e6e9a7456f13

File: Makefile.am [hunk 2]
```diff
@@ -265,6 +269,14 @@ EXTRA_DIST = \
 	ptf_tests/tests/lnt_2vm_vxlan_2nsvm_5min_ping.py \
 	ptf_tests/tests/lnt_1hotplugvm_vxlan_1nsvm.py \
 	ptf_tests/tests/config_control_port_with_link.py \
+	ptf_tests/tests/LNT_Hot_Plug_Add_Remove_multi_port.py \
+	ptf_tests/tests/LNT_Hot_Plug_Add_Remove_multiple_times.py \
+    ptf_tests/tests/LNT_ECMP_2VM_2Host_ping.py \
```
Comment (5abeel) [inline Makefile.am]: Alignment

File: Makefile.am [hunk 3]
```diff
@@ -265,6 +269,14 @@ EXTRA_DIST = \
...
+	ptf_tests/tests/LNT_ECMP_2VM_2Host_del_add.py \
+	ptf_tests/tests/LNT_ECMP_2VM_2Host_hotplug.py \
+	ptf_tests/tests/LNT_ECMP_2VM_2Host_netperf.py \
+	ptf_tests/tests/LNT_ECMP_2HotPlugVM_2Host_netperf.py \ 
+    ptf_tests/tests/LNT_ECMP_2VM_2Host_link_flap_netperf.py \
```
Comment (5abeel) [inline Makefile.am]: Alignment

File: ptf_tests/tests/LNT_ECMP_2HotPlugVM_2Host_netperf.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
+# Copyright (c) 2022 Intel Corporation.
+#
```
Comment (5abeel) [inline ptf_tests/tests/LNT_ECMP_2HotPlugVM_2Host_netperf.py:3]: Add SPDX license info line 

`# SPDX-License-Identifier: Apache-2.0`

For future reference, please add the SPDX line for all new files added.

File: ptf_tests/tests/LNT_ECMP_2VM_2Host_del_add.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
+# Copyright (c) 2022 Intel Corporation.
+#
```
Comment (5abeel) [inline ptf_tests/tests/LNT_ECMP_2VM_2Host_del_add.py:3]: Add SPDX license info.

Same comment for all other new files in this PR

**Review by 5abeel — CHANGES_REQUESTED** [review]

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-4/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/a00174c50c8d47fdb987ed4ee563137adfc43ad9

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-5/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/a00174c50c8d47fdb987ed4ee563137adfc43ad9

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-6/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/a00174c50c8d47fdb987ed4ee563137adfc43ad9

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-7/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/a00174c50c8d47fdb987ed4ee563137adfc43ad9

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-8/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/a00cb81a997b92b0a9deb82dbf8f1534376bdc0f

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-10/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/6fc984d600429aaa2fe435a5d7796e463b1b60f6

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-11/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/6fc984d600429aaa2fe435a5d7796e463b1b60f6

File: ptf_tests/common/config/LNT_ECMP_2VM_2Host_del_add.json [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    }
+  ],
+  "net_namespace": [
+    {
+      "id": "1",
+      "name": "VM0",
+      "veth_if": "veth_vm0",
+      "ip": "99.0.0.3",
+      "peer_name": "veth_host_vm0",
+      "cmds": [
```
Comment (dkaranjai) [inline ptf_tests/common/config/LNT_ECMP_2VM_2Host_del_add.json]: please dont put command strings in json, try creating them in the script

File: ptf_tests/common/config/LNT_ECMP_2VM_2Host_ping.json [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+      "mac_remote": "00:e8:ca:11:cc:02"
+    }
+  ],
+  "net_namespace": [
+    {
+      "id": "1",
+      "name": "VM0",
+      "veth_if": "veth_vm0",
+      "peer_name": "veth_host_vm0",
+      "cmds": [
```
Comment (dkaranjai) [inline ptf_tests/common/config/LNT_ECMP_2VM_2Host_ping.json]: same as above

File: ptf_tests/common/config/LNT_ECMP_2VM_2Host_netperf.json [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+      "mac_remote": "00:e8:ca:11:cc:02"
+    }
+  ],
+  "net_namespace": [
+    {
+      "id": "1",
+      "name": "VM0",
+      "veth_if": "veth_vm0",
+      "peer_name": "veth_host_vm0",
+      "cmds": [
```
Comment (dkaranjai) [inline ptf_tests/common/config/LNT_ECMP_2VM_2Host_netperf.json]: same as above

File: ptf_tests/common/utils/test_utils.py [hunk 9]
```diff
@@ -639,3 +666,244 @@ def ip_ntns_exec_ping_test(nsname, dst_ip, count="4", remote=False, hostname="",
...
+    if "Netperf version" not in result:
+        print (f"FAILED: netperf is not installed on namespace {hostname}")
+        connection.tear_down
+        return False
+    print (f"PASS: verify netperf is installed on {hostname}")
+ 
+    _, _, err= connection.execute_command("pkill -9 netserver")
+    if err:
+        print (f"FAILED: faild to pkill -9 netserver")
+        connection.tear_down
```
Comment (dkaranjai) [inline ptf_tests/common/utils/test_utils.py]: connection.tear_down()

**Review by dkaranjai — CHANGES_REQUESTED** [review]

File: ptf_tests/common/config/LNT_ECMP_2VM_2Host_netperf.json (hunk 8 above)
Comment (sjwahub) [reply ptf_tests/common/config/LNT_ECMP_2VM_2Host_netperf.json]: fixed

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-13/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/6e3f9bc3df13d2ea6aad06f80218efc638abcddc

**Comment by dkaranjai** [comment]
Hi @sjwahub Please attach regression run result after you made the changes in the json file.

**Comment by sjwahub** [comment]
> Hi @sjwahub Please attach regression run result after you made the changes in the json file.

Regression run log , [ptf_tests_run35.log](https://github.com/ipdk-io/ovs/files/9467008/ptf_tests_run35.log), in which LNT_ECMP_2VM_2Host_del_add has failure of ping without loss after deleting rule. It's not related to this change.  I'll file a sighting bug.

**Comment by sjwahub** [comment]
> 

Unit test which  failed case in suite [LNT_ECMP_2VM_2Host_del_add_unittest.txt](https://github.com/ipdk-io/ovs/files/9467876/LNT_ECMP_2VM_2Host_del_add_unittest.txt)
Test to run [ptf_tests_run35.log](https://github.com/ipdk-io/ovs/files/9467877/ptf_tests_run35.log)

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-16/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/ca26c6d777c56df186426e94fd26b95d6256b85d

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-17/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/ca26c6d777c56df186426e94fd26b95d6256b85d

**Review by dkaranjai — APPROVED** [review]
all comments addressed

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-18/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/ca26c6d777c56df186426e94fd26b95d6256b85d

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-19/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/69156be07232c48a8762b5da76eeb07d4c5ed25e

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-20/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/69156be07232c48a8762b5da76eeb07d4c5ed25e

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-21/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/69156be07232c48a8762b5da76eeb07d4c5ed25e

**Review by 5abeel — APPROVED** [review]
lgtm

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-22/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/11420156643f5509776cafec6bd2838f22d8a00e

**Comment by ipdksw** [comment]
Result:ABORTED JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-23/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/13ea1b7d46eff93eeeaedcd9a19c6709f89c7376

**Comment by ipdksw** [comment]
Result:ABORTED JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-24/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/13ea1b7d46eff93eeeaedcd9a19c6709f89c7376

**Comment by sjwahub** [comment]
0b987a9303f729090c1192150f6229b3d0ee5007

**Comment by sjwahub** [comment]
Logs 
[ptf_tests_run44.log](https://github.com/ipdk-io/ovs/files/9544709/ptf_tests_run44.log)
for   amend commit https://github.com/ipdk-io/ovs/pull/60/commits/0b987a9303f729090c1192150f6229b3d0ee5007

The message for netperf enhancement 

execute netperf -H 99.0.0.4 -l 10 -t TCP_STREAM -- -m 64 on VM0
PASS: send netperf -H 99.0.0.4 -l 10 -t TCP_STREAM -- -m 64 succeed with below output 
netperf -H 99.0.0.4 -l 10 -t TCP_STREAM -- -m 64
MIGRATED TCP STREAM TEST from 0.0.0.0 (0.0.0.0) port 0 AF_INET to 99.0.0.4 () port 0 AF_INET : demo
Recv   Send    Send                          
Socket Socket  Message  Elapsed              
Size   Size    Size     Time     Throughput  
bytes  bytes   bytes    secs.    10^6bits/sec  

131072  16384     64    10.00     644.08   
root@TRAFFICGEN:~#

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-60-26/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/dc8cb1e03311a73c75cbcb1703033f1b58cfebd1

**Pull request merged**